search/views.py

```diff
@@ -74,65 +74,339 @@ def error500(request):
+        elif expr[i] == '-' and ((len(token) > 0 and token[-1] == '(') or i == 0):
+            token.append('u-')
+            i += 1
+        elif expr[i].isdigit() or expr[i] == '.':
+            j = i
+            flot = 0
+            while j < len(expr) and (expr[j].isdigit() or expr[j] == '.'):
+                if expr[j] == '.':
+                    if flot == 0:
+                        flot = 1
+                    else:
+                        return False
+                j = j + 1
+            try:
+                val = int(expr[i:j])
+            except ValueError:
+                val = float(expr[i:j])
+            token.append(val)
+            i = j
+
+        elif expr[i] == '(' and len(token) > 0 and (str(token[-1]).isdigit() or token[-1] == ')'):
+            token.append('*')
+
+        elif expr[i] == '(' or expr[i] == ')' or expr[i] == '+' or \
+                expr[i] == '*' or expr[i] == '-' or expr[i] == '/' or \
+                expr[i] == '%' or expr[i] == '^':
+            token.append(expr[i])
+            i = i + 1
+
+        elif expr[i] == 'x' or expr[i] == 'X':
+            token.append('*')
+            i = i + 1
+
+        elif expr[i:i + 5].lower() in op_math:
+            token.append(expr[i:i + 5].lower())
+            if i + 5 >= len(expr):
+                return False
+            elif expr[i + 5] != '(':
+                return False
+            i = i + 5
+
+        elif expr[i:i + 4].lower() in op_math:
+            token.append(expr[i:i + 4].lower())
+            if i + 4 >= len(expr):
+                return False
+            elif expr[i + 4] != '(':
+                return False
+            i = i + 4
+
+        elif expr[i:i + 3].lower() in op_math:
+            token.append(expr[i:i + 3].lower())
+            if i + 3 >= len(expr):
+                return False
+            elif expr[i + 3] != '(':
+                return False
+            i = i + 3
+
+        else:
+            return False
+    return token
+
+
+def calculate(val1, val2, op):
+    if op == 'u-':
+        return - val1
+    if op == 'u+':
+        return val1
+    if op == '^':
+        return math.pow(val1, val2)
+    if op == '+':
+        return val1 + val2
+    if op == '-':
+        return val1 - val2
+    if op == '*':
+        return val1 * val2
+    if op == '/':
+        return val1 / val2
+    if op == '%':
+        return val1 % val2
+    if op == 'sin':
+        return math.sin(val1)
+    if op == 'cos':
+        return math.cos(val1)
+    if op == 'tan':
+        return math.tan(val1)
+    if op == 'cot':
+        return 1 / math.tan(val1)
+    if op == 'cosec':
+        return 1 / math.sin(val1)
+    if op == 'sec':
+        return 1 / math.cos(val1)
+    if op == 'exp':
+        return math.exp(val1)
+    if op == 'sqrt':
+        return math.sqrt(val1)
+    if op == 'log':
+        return math.sqrt(val1)
+
+
+def check_precedence(opr, i, op_precdn, op_math):
+    if opr in op_precdn and i in op_math:
```

Comment:
You could honestly shorten this function to one line using the logical operators, but that is a preference thing.